Чтобы нам как-то измерить результаты модели, реализуем метрику accuracy разделив верно предсказанные классы на размер выборки

In [1]:
def score(y_true, y_pred):
    return sum(y_true == y_pred) / len(y_true)

## Классная работа № 5

В данной классной работе мы попробуем обучить несколько алгоритмов на известном датасете.

In [2]:
from sklearn import datasets

dataset = datasets.load_wine()

In [3]:
print(dataset['DESCR'])

.. _wine_dataset:

Wine recognition dataset
------------------------

**Data Set Characteristics:**

    :Number of Instances: 178 (50 in each of three classes)
    :Number of Attributes: 13 numeric, predictive attributes and the class
    :Attribute Information:
 		- Alcohol
 		- Malic acid
 		- Ash
		- Alcalinity of ash  
 		- Magnesium
		- Total phenols
 		- Flavanoids
 		- Nonflavanoid phenols
 		- Proanthocyanins
		- Color intensity
 		- Hue
 		- OD280/OD315 of diluted wines
 		- Proline

    - class:
            - class_0
            - class_1
            - class_2
		
    :Summary Statistics:
    
    ============================= ==== ===== ======= =====
                                   Min   Max   Mean     SD
    ============================= ==== ===== ======= =====
    Alcohol:                      11.0  14.8    13.0   0.8
    Malic Acid:                   0.74  5.80    2.34  1.12
    Ash:                          1.36  3.23    2.36  0.27
    Alcalinity of Ash:            1

Согласно описанию датасета перед нами выборка 178 винных напитков, для которых мы будем предсказывать один из 3 классов.\
Для начала разделим выборку на обучающую и отложенную.

In [4]:
from sklearn.model_selection import train_test_split

X = dataset.data
y = dataset.target
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=17)

Ссылка на Google форму для проверки ответов: https://forms.gle/MfT4ZSibVKHgHDKi8

**1. Обучите на имеющейся выборке случайный лес (RandomForestClassifier) с деревьями максимальной глубины 1. Найдите точность (определяется с помощью функции `score`) у этого алгоритма.**\
Используйте параметр `random_state=17` для воспроизводимости результатов.

In [5]:
# Ваш код здесь

**2. Обучите на имеющейся выборке случайный лес (RandomForestClassifier) с деревьями максимальной глубины 2. Найдите точность (определяется с помощью функции `score`) у этого алгоритма.**\
Используйте параметр `random_state=17` для воспроизводимости результатов.

In [7]:
# Ваш код здесь

**3. С помощью `GridSearchCV` найдите наилучшие параметры из перечисленных ниже.**\
Используйте параметр `random_state=17` для воспроизводимости результатов.\
Максимальная глубина ограничена 1 у всех деревьев (параметр `max_depth` должен быть равен 1).

In [9]:
# params for grid search
rf_params = {'min_samples_leaf': list(range(1, 5)),
             'n_estimators': [1, 10, 20, 50, 100]}

# Ваш код здесь

**4. Какой OOB score у модели с параметрами, подобранными в задании 3?**\
Используйте параметр `random_state=17` для воспроизводимости результатов.

In [12]:
# Ваш код здесь

**5. Какой признак (порядковый номер) обладает наибольшей важностью у модели с параметрами, подобранными в задании 3?**\
Используйте параметр `random_state=17` для воспроизводимости результатов.\
Для оценки важности надо воспользоваться impurity-based подходом.

In [14]:
# Ваш код здесь

**6. Какой точностью обладает алгоритм бэггинга на деревьях решений с максимальной глубиной 2?**\
Используйте параметр `random_state=17` для воспроизводимости результатов.

In [16]:
# Ваш код здесь